Render Effect Hooks For Vanilla Coloring & Custom Effect Layers

RANKSHANK opened the pull request:
![screenshot 2016-02-11 10 06 06](https://cloud.githubusercontent.com/assets/1190024/12964981/21bb184c-d0a7-11e5-9321-e1e680e083a3.png)
## Overview
This opens up the effect rendering to modders, allowing for item stack and armor stack aware coloring of the enchantment overlay respectively using added methods in the Item.class. This allows for potential visual feedback of item/armor status through glint color or simple vanity. Modders can take this a step further and entirely rework the overlay to their needs. Different texture, different color, different speed etc.

#### Included is a test mod that does the following:
Adds an ItemArmor(lazily as possible) that has in meta = 0 an item with a custom EffectPassHandler for its stack and armor appearance, and for meta = 1 a different dual glint color for its ItemStack in comparison to its Armor Model (leather helmet) 

#### Patch changes are as follows:
1. RenderItem:
    * Replaces vanilla effect overlay with EffectPassHandler#ItemStackPassHandler
2. LayerArmorBase
    * Replaces vanilla effect overlay with EffectPassHandler#ArmorPassHandler
3. Item
    * Adds constant int reflecting the vanilla glint value.
    * Adds method for returning glint color for itemstacks and armor models respectively
    * Adds method for returning alternate EffectPassHandler for itemstacks and armor models respectively

#### Potential Pitfalls:
1. The integers passed for the glint require the a byte dedicated to the alpha channel. This may cause confusion for modders given that most vanilla methods that take colors in the form of integers do not have alpha channels resulting in the appearance that the effect is not being rendered when the alpha channel is empty. This can be remedied simply with : [0xff << 24 | alphalessInteger]
2. The EffectPassHandler is meant to be treated as a singleton. Instantiating a new one per call is just a potential for more performance degradation.

Spartan322 commented:
Pretty sure the javadocs in Forge aren't suppose to have author tags iirc, also the glintValue's assignment in renderItemGlint needs one more space for consistency

williewillus commented:
Wow an actually well documented and useful pr :P

+1

williewillus commented:
On another note - I don't think forge-added classes and methods need to have sideonly client, unless it actually touches net.minecraft.client code
Someone check me on that though

Spartan322 commented:
Pretty sure the only reason the SideOnly annotation is even there is for vanilla/continuing compatibility.

RANKSHANK commented:
Yeah I'm not 100% sure where the SideOnly stands on the item class. When I find out I'll patch it out if necessary with the event classes one since only GuiScreenEvent has that annotation :P

heldplayer commented on patches/minecraft/net/minecraft/item/Item.java.patch:
```diff
@@ -723,12 +723,41 @@
++     * Vanilla's default glint value
++     * */
++    @SideOnly(Side.CLIENT)
++    private static final int defaultGlint = -8372020;
```
Might be better to put this constant in hexadecimal notation rather than int so you can actually get an understanding of what's going on.

Example: 0xFF8040CC

RANKSHANK replied:
@heldplayer it's now instantiated with hex, thanks for the tip!

kashike commented on patches/minecraft/net/minecraft/item/Item.java.patch:
```diff
@@ -723,12 +723,41 @@
++    @SideOnly(Side.CLIENT)
++    public int getArmorGlintColor(ItemStack stack, EntityLivingBase wearer, int armorSlot)
++    {
++        return  defaultGlint;
```
One too many spaces here

kashike commented on src/main/java/net/minecraftforge/client/ForgeHooksClient.java:
```diff
@@ -578,4 +580,37 @@ public static void fillNormal(int[] faceData, EnumFacing facing)
+       OverlayGlintEvent.OverlayItemStackEvent event = new OverlayGlintEvent.OverlayItemStackEvent(stack, model);
+       if(!MinecraftForge.EVENT_BUS.post(event))
+       {
+    	   glintValue = (event.glintValue);
```
Formatting is off

RANKSHANK replied:
@kashike thanks for pointing that out!

RANKSHANK commented:
Fixed up some horrible formatting on my part. Refactored the ItemStack event's name for consistency's sake & made a few tweaks for readability and such. Thanks for the help guys

LexManos commented on patches/minecraft/net/minecraft/client/renderer/entity/RenderItem.java.patch:
```diff
@@ -9,6 +9,32 @@
+         GlStateManager.func_179109_b(f, 0.0F, 0.0F);
+         GlStateManager.func_179114_b(-50.0F, 0.0F, 0.0F, 1.0F);
+-        this.func_175035_a(p_180451_1_, -8372020);
++        this.func_175035_a(p_180451_1_, net.minecraftforge.client.ForgeHooksClient.getGlintColor());
```
It would probably be better suited to pass the ItemStack down here as well, yes it's not available in this function but as this is a private function you can pass the value in and keep the patch small with the proper exc/srg lines.

LexManos commented on patches/minecraft/net/minecraft/client/renderer/entity/layers/LayerArmorBase.java.patch:
```diff
@@ -37,11 +37,21 @@
+             GlStateManager.func_179112_b(768, 1);
+-            float f2 = 0.76F;
+-            GlStateManager.func_179131_c(0.5F * f2, 0.25F * f2, 0.8F * f2, 1.0F);
++            net.minecraftforge.client.ForgeHooksClient.applyGlintColor();
```
Same goes for here, both methods are private and only referenced one time. Modders shouldn't be accessing private methods so it should be safe to break binary compatibility in this instance.

LexManos commented on src/main/java/net/minecraftforge/client/ForgeHooksClient.java:
```diff
@@ -578,4 +580,37 @@ public static void fillNormal(int[] faceData, EnumFacing facing)
+    {
+        OverlayGlintEvent.ArmorOverlayEvent event = new OverlayGlintEvent.ArmorOverlayEvent(stack, model, entity, slot);
+        if(!MinecraftForge.EVENT_BUS.post(event))
+        {
```
if (

LexManos commented on src/main/java/net/minecraftforge/client/ForgeHooksClient.java:
```diff
@@ -578,4 +580,37 @@ public static void fillNormal(int[] faceData, EnumFacing facing)
+        if(!MinecraftForge.EVENT_BUS.post(event))
+        {
+            glintValue = event.glintValue;
+            return true;
```
Passing in the itemstack would prevent the need for this potentially thread-unsafe static variable here.

LexManos commented on src/main/java/net/minecraftforge/client/event/OverlayGlintEvent.java:
```diff
@@ -0,0 +1,87 @@
+ * Events for hooking into the Glint effect Rendering<br>
+ * Will allow for adding effects to items, creating custom overlays, modifying vanilla
+ * colors, and canceling of glint overlays. 
+ * @author RANKSHANK
```
No

LexManos commented on src/main/java/net/minecraftforge/client/event/OverlayGlintEvent.java:
```diff
@@ -0,0 +1,87 @@
+     * {@link ModelBase} during {@link ArmorOverlayEvent}<br>
+     * */
+    public final T theModel;
+
```
Also it is now becoming the standard to go getters and setters instead of final fields here. It allows for better forward flexibility. So please follow that format.

LexManos commented on src/main/java/net/minecraftforge/client/event/OverlayGlintEvent.java:
```diff
@@ -0,0 +1,87 @@
+ * @author RANKSHANK
+ */
+public class OverlayGlintEvent<T> extends Event
+{
```
Generics can screw things up, make sure to mark this as an abstract class so that modders do not instantiate it itself.

LexManos commented on src/main/java/net/minecraftforge/client/event/OverlayGlintEvent.java:
```diff
@@ -0,0 +1,87 @@
+    public int glintValue;
+
+    public OverlayGlintEvent(ItemStack stack, T model, int color)
+    {
```
Private constructor.

RANKSHANK commented:
![screenshot 2016-02-15 03 08 53](https://cloud.githubusercontent.com/assets/1190024/13035000/7a75e7a2-d398-11e5-9267-695b66f72a3b.png)
Almost a complete overhaul of the system. Moved to encapsulating the vanilla effect passes in EffectPassHandler classes. Works via pre & post model render calls. This has allowed for Events to be free of model handles and for effect coloring/modeling/texturing to work in a pass based system. Test mod has also been updated to reflect the changes, mainly ItemSwords with fire aspect now have a new overlay more akin to flames.

@LexManos Thanks for all the help- I've hopefully fixed what you've mentioned in the rewrite process. Definitely the potential for my tard ass making a few more errors in the rewrite

RainWarrior commented on patches/minecraft/net/minecraft/client/renderer/entity/RenderItem.java.patch:
```diff
@@ -9,7 +9,61 @@
+-    private void func_180451_a(IBakedModel p_180451_1_)
++    private void func_180451_a(IBakedModel p_180451_1_, ItemStack stack)
+     {
+-        GlStateManager.func_179132_a(false);
```
Allowing customization of the GL state here breaks anything trying to render items faster than the normal GUI does, like JEI. I'd go for more limited functionality - allow changing color + glint texture only - then the batching should still be possible.

RainWarrior replied:
This comment is still relevant.

RANKSHANK replied:
@RainWarrior Sorry I just want to make sure it's worth ripping out custom effect support as there's little to no reason to include it when it'll have to follow the vanilla glint's translation and scaling

I've only found [model type batching] (https://github.com/mezz/JustEnoughItems/blob/1.8.9/src/main/java/mezz/jei/gui/ingredients/GuiItemStackFastList.java#L110)... and then the overlay is run through afterwards which is the text information... the state is [still being flipped](https://github.com/mezz/JustEnoughItems/blob/1.8.9/src/main/java/mezz/jei/gui/ingredients/GuiItemStackFast.java#L170) during the model rendering for the effect passes since they're all handled in the one method... at least from what I've seen, aside from the batched lighting calls, which can be moved out from the passes, it is virtually identical to [my approach](https://github.com/MinecraftForge/MinecraftForge/pull/2458/files#diff-e3a7ac5de7307337f8f7a2fd797d506aR166)

Sorry @mezz I hope you don't mind me asking for some input on the matter from you

With JEI for supporting this you'd end up putting
```java
if (itemStack.hasEffect()) {
    EffectPassHandler handler = itemStack.getItem.getItemEffectRenderer(itemStack);
    if(handler.isEssential()){
        handler.bindTexture
        //run the passes
    }
```
In the renderEffect method to allow for the custom effects to be run through instead. Totally an optional thing to do since JEI already has its own included method that it uses for effect rendering.

The only time I can see that JEI would run into the patched calls with the EffectPassHandler support is included is when the it advocates to 'renderSlowly'(via vanilla)... unless I am missing something?

mezz replied:
The original goal was to render the whole item list in one shot basically, caching a list of quads that could be dumped to the GPU in one call. Once it was proven, Forge could do something similar maybe. I ended up stopping short of that goal because of hangups like this, where special cases allow modders to interject their own OpenGL calls in the middle of the rendering pipeline, or rely on other things that can't be cached. For good rendering performance, I think this should be kept as simple as possible.

While it is possible for me to handle this case, I don't think it's a step in the right direction. Adding more power for modders is good, but it should be weighed against the costs on the other side like the ability to reason about and optimize the rendering code.

RANKSHANK replied:
Fair enough. I'm no where near the level of being able to write something that'll account for all possible calls to render items in the pipeline, cache them, and efficiently batch the lot of them together for the single draw call, so I'll definitely be leaving things like this to the more knowledgeable.

mezz replied:
Well, honestly nobody is at that level or it would be done in JEI heh.
Don't be afraid to dig into it more.

RANKSHANK replied:
Ehh I've really got little reason to dig more haha my knowledge of rendering pipelines is next to nil :P I was just pushing this for more modding power with the item/ armor rendering. IE being able to add enchantment shaders and such as opposed to boosting vanilla performance

RainWarrior commented on src/main/java/net/minecraftforge/client/ForgeHooksClient.java:
```diff
@@ -578,4 +582,21 @@ public static void fillNormal(int[] faceData, EnumFacing facing)
+    public static ItemStackEffectPassHandler getStackGlintEffectPassHandler(ItemStack stack)
+    {
+        EffectOverlayEvent.ItemStackEffectOverlayEvent event = new EffectOverlayEvent.ItemStackEffectOverlayEvent(stack);
+        MinecraftForge.EVENT_BUS.post(event);
```
Firing an event for each item each frame is probably not a good idea - enchanting glint puts the noticeable dent in the FPS already.

RainWarrior commented on src/main/java/net/minecraftforge/client/model/pipeline/EffectPassHandler.java:
```diff
@@ -0,0 +1,289 @@
+         * @param pass The effect pass
+         * @param partialTicks The partialTick value passed to the armor renderer
+         * */
+        public abstract void prePass(ItemStack stack, EntityLivingBase wearer, @SuppressWarnings("rawtypes") RendererLivingEntity renderer, int slot, int pass, float partialTicks);
```
No raw types please. This is 2016.

RANKSHANK commented:
![screenshot 2016-02-15 19 41 33](https://cloud.githubusercontent.com/assets/1190024/13043833/d1385bfc-d41e-11e5-9d5a-6fae40e7ea77.png)

Thanks @RainWarrior 

I've entirely removed the events from the pull.

The EffectPassRenderer is now delegated to an Item hook.

Did some nano runtime testing v.s. clean to make improvements to the encapsulated translations. Nano time during areas where many models are rendered with vanilla effect (such as the brewing tab) come to be almost twice as fast during renderEffect. Armor seems to be relatively similar (both vanilla and revamped have similar sway values in nanos)

As for batching I'm thinking on that... perhaps changing the renderEffect method to allocate the EffectPassHandler so a fast variant can be fed to the renderEffect calls. I'm hesitant to manually do the effectColor calls directly in the vanilla renderEffect method because it removes a lot of the power over custom effect appearances, like the glintTest item's EmberEffectHandlers that give the example of a different overlay.

Added pitfall: instantiating EffectPassHandlers in the hook since they are designed to be singletons and will cause a lot of lag related issues.

RANKSHANK commented:
Changes:
1. Added EffectPassHandler#bindTexture and removed the T#textureBind access from the pass methods to discourage state changes during passes
2. Added required super(boolean) to the constructor to allocate a return for EffectPassHandler#isEssential, so that modders can check if the handler is necessary
3. Removed model replacement methods. It wasn't working as intended and thus was creating clutter. 

I'm hoping that this update will improve the 'batchability' of the pass handlers. Now you can call bindTexture once and run the passes over as many models as necessary. isEssential() should allow for skipping of the effect passes  when that extra bit of performance is needed. Sniffing around the JEI git brought my hopes up a bit for ulterior usage cases. since the actual pass methods are pretty light in terms of code and the EffectPassHandlers are not only non item dependent(interchangable with faster pass handlers) but also have public availability
